### PR TITLE
Add support for handling DTLS alerts

### LINKED ISSUE #8: Handle bye packet
In videoroom demo, when closing a browser we sometimes get a packet like:

```
<<21, 254, 253, 0, 1, 0, 0, 0, 0, 0, 1, 0, 26, 0, 1, 0, 0, 0, 0, 0, 1, 86, 179, 186, 47, 158, 179, 56, 119, 185, 211, 192, 255, 71, 253, 59, 63, 122, 131>>
```

or

```
<<21, 254, 253, 0, 1, 0, 0, 0, 0, 0, 1, 0, 26, 0, 1, 0, 0, 0, 0, 0, 1, 2, 251, 29, 34, 184, 32, 88, 215, 151, 16, 84, 104, 5, 58, 178, 238, 113, 220>>
```

that make the SRTP decryptor crash. It seems that they are neither RTP nor RTCP packets, so they may be kind of DTLS bye packets. We should make sure that no DTLS packets are forwarded with the media stream.

### PR DESCRIPTION
**Pull request by mickel8**
* closes #8 
* allows for receiving regular DTLS packets

### CODE REVIEW

File: c_src/ex_dtls/native.c [hunk 1]
```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
...
+    UNIFEX_TERM res_term = process_result_ok(state->env, state, packets);
+    unifex_payload_release(packets);
+    return res_term;
+  }
 
-      DEBUG("WBIO: read: %d bytes", read_bytes);
+  // handle READ errors including DTLS alerts
+  int error = SSL_get_error(state->ssl, ret);
+  DEBUG("SSL ERROR: %d", error);
+  return process_result_error(state->env, error);
```
Comment (mickel8) [inline c_src/ex_dtls/native.c:159]: I am not sure how to handle it. We can:
* raise here
* return error with error value but error value will mean nothing for the user
* return :ok and if DTLS alert occurred we can send message from `ssl_info_cb`

`error` can mean DTLS alerts but also some other retryable or non retryable errors. I would leave it as it is and see how it behaves.

Comment (mat-hek) [reply c_src/ex_dtls/native.c:159]: In general, we should raise on critical, non-retryable errors, warn on non-critical errors and retry on the retryable ones ;) If we don't know which errors are critical, let's always warn, at least for now.

File: c_src/ex_dtls/native.c [hunk 2]
```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
...
       unifex_payload_release(gen_packets);
       unifex_payload_release(client_keying_material);
       unifex_payload_release(server_keying_material);
       return res_term;
-    default:
-      DEBUG("SSL ERROR: %d", res);
-      return unifex_raise(env, "Handshake failed: SSL error");
     }
+  } else {
+    // we have no pending data to send - send empty binary
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:212]: Can't we just send nothing?

File: c_src/ex_dtls/native.c [hunk 3]
```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
...
+    UnifexPayload *gen_packets = (UnifexPayload *)unifex_payload_alloc(
+        state->env, UNIFEX_PAYLOAD_BINARY, 0);
+    state->hsk_finished = 1;
+    UNIFEX_TERM res_term = process_result_hsk_finished(
+        state->env, state, client_keying_material, server_keying_material,
+        keying_material->protection_profile, gen_packets);
+    unifex_payload_release(gen_packets);
+    unifex_payload_release(client_keying_material);
+    unifex_payload_release(server_keying_material);
+    return res_term;
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:222]: this code seems to be a duplicate from a few lines above, let's reorganize that

File: c_src/ex_dtls/native.c [hunk 4]
```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
...
+        free(pending_data);
+        unifex_payload_release(gen_packets);
+        return res_term;
+      }
+    } else {
+      return process_result_hsk_want_read(state->env);
+    }
+  default:
+    DEBUG("SSL ERROR: %d", ssl_error);
+    return process_result_error(state->env, ssl_error);
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:260]: Quite a lot of nesting here. Maybe using `goto` to handle errors would help

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: c_src/ex_dtls/native.c [hunk 5]
```diff
@@ -6,7 +6,10 @@
 #include "dyn_buff.h"
 #include "native.h"
 
-#define BUF_LEN 2048
+void ssl_info_cb(const SSL *ssl, int where, int ret);
+UNIFEX_TERM handle_regular_read(State *state, char data[], int ret);
+UNIFEX_TERM handle_handshake_inprogress(State *state, int ret);
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:11]:
```suggestion
UNIFEX_TERM handle_handshake_in_progress(State *state, int ret);
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: c_src/ex_dtls/native.c (hunk 2 above)
Comment (mickel8) [reply c_src/ex_dtls/native.c]: I think that after refactoring it looks quite good. We got rid of this else [here](https://github.com/membraneframework/ex_dtls/blob/dtls-alert/c_src/ex_dtls/native.c#L188-L196).
Now we always allocate payload for pending data. When pending data is > 0 we copy it to payload. In other case we just return empty payload. Sending nothing would require additional return type in `native.spec.exs` e.g. `handshake_finished_with_packets`.

File: c_src/ex_dtls/native.c [hunk 6]
```diff
@@ -182,39 +195,27 @@ UNIFEX_TERM handle_handshake_finished(State *state) {
...
-      return unifex_raise(state->env, "Handshake failed: write BIO error");
-    } else {
-      DEBUG("WBIO: read: %d bytes", read_bytes);
-      gen_packets = (UnifexPayload *)unifex_payload_alloc(
-          state->env, UNIFEX_PAYLOAD_BINARY, pending_data_len);
-      memcpy(gen_packets->data, pending_data, pending_data_len);
-      gen_packets->size = (unsigned int)pending_data_len;
-      free(pending_data);
+    if (read_pending_data(gen_packets, pending_data_len, state) < 0) {
+      unifex_payload_release(gen_packets);
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:202]: you're already releasing that in cleanup

File: c_src/ex_dtls/native.c [hunk 7]
```diff
@@ -223,35 +224,42 @@ UNIFEX_TERM handle_handshake_inprogress(State *state, int ret) {
...
+  BIO *wbio = SSL_get_wbio(state->ssl);
+  int read_bytes = BIO_read(wbio, pending_data, pending_data_len);
+  if (read_bytes <= 0) {
+    DEBUG("WBIO: read error");
+    free(pending_data);
+  } else {
+    DEBUG("WBIO: read: %d bytes", read_bytes);
+    memcpy(gen_packets->data, pending_data, pending_data_len);
+    gen_packets->size = (unsigned int)pending_data_len;
+    free(pending_data);
```
Comment (mat-hek) [inline c_src/ex_dtls/native.c:257]: move out of `if`

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Review by mat-hek — APPROVED** [review]

**Pull request merged**